Dual Stack support for localnet and nodeport services

gvbalaji opened the pull request:
Dual Stack support for localnet and nodeport services

    1. Added dualstack support for localnet and nodeport service
        a. Now we install base localnet rules for both IPv4 and IPv6
        b. For Nodeport services, we install IP rules based on IP family of cluster IP
        c. Gateway NAT rules for both IPv4 and IPv6
        d. L3gateway setup considers both IPv4 and IPv6
    2. Some low hanging refactring for unit testing
    3. Integration tests for IPv6 and dual-stack
    4. Started unit tests for gateway-localnet.go. Coverage stands at 77% now.
    

@danwinship Please let me know your feedback.

danwinship commented:
You need to rebase and get rid of my annotation patch.

Please split out the `LinkNeighAdd` fix into its own separate commit.

There are lots of code style issues; run `gofmt -s -w go-controller/pkg go-controller/cmd` to fix them all up, and then look over the changes it made to get a sense of how go code is supposed to look. (There ought to be some way to get your IDE to run gofmt automatically?)

I know I agreed that the existing code was messy and needed to be refactored, but I guess I shouldn't have said that, and you should avoid refactoring until you have more of a sense of golang coding style, because I think a lot of the refactoring here is not really right. I think this PR should be a lot smaller. But anyway, do the other stuff above and re-push and I'll re-review.

gvbalaji commented:
Thanks @danwinship for the comments. Will make the changes you suggested. I did run the format in my Goland ide not sure why it didn't work.  Main refactoring I've done is is break down larger functions into smaller ones but we can discuss that.

gvbalaji commented:
@danwinship I've removen net_linux.go changes from here and created a separate PR as you suggested. For gateway_localnet.go changes I would suggest to take a look at the file view once.

gvbalaji commented:
Added unit test cases for localnet gateway . This is the beginning of that for this file which was possible due to refactoring.

Added integration tests for IPv6 and dual stack for localnet along with IPv4

danwinship commented on go-controller/pkg/node/gateway_init.go:
```diff
@@ -103,7 +103,8 @@ func (n *OvnNode) initGateway(subnet *net.IPNet, nodeAnnotator kube.Annotator,
 		}
-		prFn, err = n.initSharedGateway(subnet, gatewayNextHop, gatewayIntf, nodeAnnotator)
+		//FIXME DUAL-STACK
+		prFn, err = n.initSharedGateway(subnets[0], gatewayNextHop, gatewayIntf, nodeAnnotator)
```
`initSharedGateway` doesn't even look at its `subnet` argument, so you can just remove it and not need a FIXME.

gvbalaji replied:
Will do.

danwinship commented on go-controller/pkg/node/gateway_localnet.go:
```diff
@@ -3,7 +3,9 @@
 package node
 
 import (
+	//"debug/elf"
```
leftover from debugging?

gvbalaji replied:
It was there before and hence I did not remove it. I suppose I can remove it.

danwinship commented on go-controller/pkg/node/gateway_localnet.go:
```diff
@@ -3,7 +3,9 @@
 import (
+	//"debug/elf"
 	"fmt"
+	"github.com/vishvananda/netlink"
```
This should be below with the other github imports.

Well... the `"k8s.io/klog"` import should be moved to be with the other k8s.io imports, and then the netlink import would go at the end of the github imports where the klog import currently is...

gvbalaji replied:
Will do. I thought go formatting will fix all these which I run before committing.

danwinship commented on go-controller/pkg/node/gateway_localnet.go:
```diff
@@ -116,32 +122,107 @@ func generateGatewayNATRules(ifname string, ip net.IP) []iptRule {
-func initLocalnetGateway(nodeName string, subnet *net.IPNet, wf *factory.WatchFactory, nodeAnnotator kube.Annotator) error {
-	// Create a localnet OVS bridge.
+func initLocalnetGateway(nodeName string, subnets []*net.IPNet, wf *factory.WatchFactory, nodeAnnotator kube.Annotator) error {
+
```
(Don't start functions with a blank line. Except when the function header is wrapped onto a second line and gets indented and so it's hard to read without adding a blank line.)

gvbalaji replied:
Will do. I thought go formatting will fix all these which I run before committing.

danwinship replied:
The "where blank lines do and don't belong" rules are kind of vague and not universally consistent, so gofmt doesn't get involved with that.

Likewise for the import ordering; there are a few different styles. Some people split "local" (same-project) imports out into a separate group, some don't, etc.

danwinship commented on go-controller/pkg/node/gateway_localnet.go:
```diff
@@ -116,32 +122,107 @@ func generateGatewayNATRules(ifname string, ip net.IP) []iptRule {
-		localnetBridgeName)
+
+	// Evaluate input subnets
+	err := evaluateInputNetworks(subnets)
```
This comment doesn't provide any information beyond what the function name says. It should be removed.

But also, you don't need to validate the data anyway. That gets done by higher levels. You can assume that the data that was passed to you here is a supported configuration.

gvbalaji replied:
Will remove this method and tests.

danwinship commented on go-controller/pkg/node/gateway_localnet.go:
```diff
@@ -116,32 +122,107 @@ func generateGatewayNATRules(ifname string, ip net.IP) []iptRule {
 
-	ifaceID, macAddress, err := bridgedGatewayNodeSetup(nodeName, localnetBridgeName, localnetBridgeName, true)
+	// Setup local brifge and ports
+	ifaceID, macAddress, link, err := createAndSetupLocalBridgePorts(localnetBridgeName, nodeName)
```
again, comment doesn't add anything. Basically, most of the added comments should be removed

trozet replied:
The comment still has spelling errors, and isn't needed. I don't think this change is adding dual-stack support. If this is clean up then move it to its own commit.

danwinship commented on go-controller/pkg/node/gateway_localnet.go:
```diff
@@ -116,32 +122,107 @@ func generateGatewayNATRules(ifname string, ip net.IP) []iptRule {
 
+	// Set the link with the addresses from IP object
+
+	err = addLinkAddresses(link, ipData)
```
I think a lot of these helper functions make the code _less_ clear, because now I need to go look somewhere else to figure out exactly what it's doing. I would say for now, please un-split-out all of the functions you split out and just leave `initLocalnetGateway` big and sprawling. There are cleanups that could be made here, but I don't think the refactoring you're doing is in the spirit of golang coding style.

gvbalaji replied:
I think we need to talk about this one. In my understanding there is nothing is that is not idiomatic go and I believe (respectfully)  we are better off with functions with clear responsibilities especially considering what the main method would have become with the dual stack support.

girishmg replied:
@gvbalaji I agree with @danwinship here. For example: this function that you have added

```
func linkNeighExists(link netlink.Link, gatewayIP net.IP, macAddress net.HardwareAddr) bool {
	exists, err := util.LinkNeighExists(link, gatewayIP, macAddress)
	if err != nil {
		klog.Errorf("Error in exists call for MAC binding for %s on %s: %v. Will try to add the binding.",
			gatewayIP, localnetGatewayNextHopPort, err)
	}
	return exists
}
```

wraps the function just for logging an error. Why can't one directly use util.LinkNeighExists() instead of this wrapper?

gvbalaji replied:
I could have. I did it this way because we are consciously ignoring the error if anything happens there and logging a message saying we will still go ahead with add to give us one more chance at it. Hence I thought it's a small unit by itself. Now if one wants to just unit test this method for both cases we could ( However I just started unit testing on this file. Yet to figure out how to make UT widely applicable  considering the file uses its dependencies all over the place)

gvbalaji replied:
I made some pointed changes in this regard and resolving this.

danwinship commented on go-controller/pkg/node/gateway_localnet.go:
```diff
@@ -116,32 +122,107 @@ func generateGatewayNATRules(ifname string, ip net.IP) []iptRule {
-func localnetGatewayNAT(ipt util.IPTablesHelper, ifname string, ip net.IP) error {
-	rules := generateGatewayNATRules(ifname, ip)
-	return addIptRules(ipt, rules)
+func localnetGatewayNAT(localNetdata []*localnetData, ifname string) error {
```
(should probably either move the definition of `localnetData` to before here, or else move this function to after the definition of `localnetData`)

danwinship commented on go-controller/pkg/node/gateway_localnet.go:
```diff
@@ -151,92 +232,185 @@ func initLocalnetGateway(nodeName string, subnet *net.IPNet, wf *factory.WatchFa
 
-	if config.Gateway.NodeportEnable {
-		err = localnetNodePortWatcher(ipt, wf, gatewayIP)
+func newIPV4Data() (*localnetData, error) {
```
`newIPV4Data` and `newIPV6Data` can just be merged into `constructIPData`

gvbalaji replied:
Since for each IP family there are separate IP address, next hop and IPT objects , wanted to separate them this way so that it's clear and easier on the eyes and also easier to write a unit test on the method when we get there. I would prefer to keep it this way if that's okay with you. I'll address the inconsistencies in the naming conventions as per your suggestion.

danwinship commented on go-controller/pkg/node/gateway_localnet.go:
```diff
@@ -151,92 +232,185 @@ func initLocalnetGateway(nodeName string, subnet *net.IPNet, wf *factory.WatchFa
-	err = localnetGatewayNAT(ipt, localnetGatewayNextHopPort, gatewayIP)
-	if err != nil {
-		return fmt.Errorf("Failed to add NAT rules for localnet gateway (%v)", err)
+func constructIPData(subnets []*net.IPNet) ([]*localnetData, error) {
```
naming inconsistency about "IPData" vs "localnetData"...

gvbalaji replied:
Will correct this.

danwinship commented on go-controller/pkg/node/gateway_localnet.go:
```diff
@@ -151,92 +232,185 @@ func initLocalnetGateway(nodeName string, subnet *net.IPNet, wf *factory.WatchFa
+		gatewayNextHop:     gatewayNextHop,
+		gatewaySubnetMask:  gatewaySubnetMask,
+		gatewayIPCIDR:      &net.IPNet{IP: gatewayIP, Mask: gatewaySubnetMask},
+		gatewayNextHopCIDR: &net.IPNet{IP: gatewayNextHop, Mask: gatewaySubnetMask},
```
`gatewayIPCIDR` and `gatewayNextHopCIDR` are each only used in one place, so they could be removed from `localnetData` and you could just put the `&net.IPNet{...}` definition inline where they're needed

gvbalaji replied:
Will look into this.

danwinship commented on go-controller/pkg/node/gateway_localnet.go:
```diff
@@ -270,57 +444,78 @@ func localnetIptRules(svc *kapi.Service, gatewayIP string) []iptRule {
+	}
+	for _, lnData := range localNetdata {
+		if lnData.ipVersion == ipVersion {
+			return lnData
```
it seems like it would be simpler to just have

    type localnetNodePortWatcherData struct {
            ipv4Data *localnetData
            ipv6Data *localnetData
    }

gvbalaji replied:
To me it's actually a bit simpler the way I implemented . Because based on the configuration either one OR both have a valid value and the other one nil. Also we need only one of them for nodeport service but need both of them for base rules. Hence having them in a slice makes it easier for all the scenarios. Just loop through them  and use the right one based on context throughout.  I would prefer to keep it this way.

danwinship commented on go-controller/pkg/node/gateway_localnet.go:
```diff
@@ -270,57 +444,78 @@ func localnetIptRules(svc *kapi.Service, gatewayIP string) []iptRule {
+	return err
+}
+
+func addServiuceHandler(wf *factory.WatchFactory, npw *localnetNodePortWatcherData) error {
```
typo "Serviuce"

gvbalaji replied:
Will correct this.

gvbalaji commented:
Thanks @danwinship for the comments. I replied to your comments above and  I'll work on the fixes.

gvbalaji commented:
I have either answered or addressed all the comments. I have resolved all the comments except one about un-splitting the methods and left a reply there.

gvbalaji commented:
I added many unit tests to gateway-localnet.go which didn't have any. We have a very good coverage now for this file.

gvbalaji commented:
The coverage for gateway-localnet.go is 76.5% at this point :) 
Most of the non-covered areas are log statements and some error conditions.

To see the coverage do the following:
1. Checkout this PR
2. Goto go-controller/pkg/node
3. Run this command  "sudo go test -coverprofile cp.out"
4. Run this command "go tool cover -html=cp.out"
5. View the html file in a browser and from the drop down menu on the top select  gateway-localnet.go file

gvbalaji commented:
I should have squashed those commits @trozet as they were related which I did now. Sorry about the confusion.

gvbalaji commented on go-controller/pkg/node/gateway_init.go:
```diff
@@ -103,7 +103,8 @@ func (n *OvnNode) initGateway(subnet *net.IPNet, nodeAnnotator kube.Annotator,
 		}
-		prFn, err = n.initSharedGateway(subnet, gatewayNextHop, gatewayIntf, nodeAnnotator)
+		//FIXME DUAL-STACK
+		prFn, err = n.initSharedGateway(subnets[0], gatewayNextHop, gatewayIntf, nodeAnnotator)
```
Done

dcbw commented:
@gvbalaji I think this PR was obsoleted by another one? Can we close it?

gvbalaji commented:
Sure. I'll close this. This has been addressed in other PRs.
